src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,374 @@
+     * @return {@code true} if the restriction operator is one of the contains operations, {@code false} otherwise.
+     */
+    public boolean isContains()
+    {
+        return operator == Operator.CONTAINS
+               || operator == Operator.CONTAINS_KEY
+               || columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT;
+    }
+
+    @Override
+    public boolean needsFilteringOrIndexing()
+    {
+        // The need for filtering or indexing is a combination of columns expression type and operator
+        // Therefore, we have to take both into account.
+        return columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT
+               || operator.requiresFilteringOrIndexingFor(columnsExpression.columnsKind());
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions)
+    {
+        columnsExpression.addFunctionsTo(functions);
+        values.addFunctionsTo(functions);
+    }
+
+    @Override
+    public boolean needsFiltering(Index.Group indexGroup)
+    {
+        for (ColumnMetadata column : columns())
+        {
+            if (!isSupportedBy(indexGroup.getIndexes(), column))
+                return true;
+        }
+        return false;
+    }
+
+    private boolean isSupportedBy(Iterable<Index> indexes, ColumnMetadata column)
+    {
+        if (isOnToken())
+            return false;
+
+        for (Index index : indexes)
+        {
+            if (index.supportsExpression(column, operator))
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Index findSupportingIndex(Iterable<Index> indexes)
+    {
+        if (isOnToken())
+            return null;
+
+        for (Index index : indexes)
+            if (isSupportedBy(index))
+                return index;
+        return null;
+    }
+
+    @Override
+    public boolean isSupportedBy(Index index)
+    {
+        if (isOnToken())
+            return false;
+
+        for (ColumnMetadata column : columns())
+        {
+            if (index.supportsExpression(column, operator))
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public List<ClusteringElements> values(QueryOptions options)
+    {
+        assert operator == Operator.EQ || operator == Operator.IN || operator == Operator.ANN;
+        return bindAndGetClusteringElements(options);
+    }
+
+    @Override
+    public RangeSet<ClusteringElements> restrict(RangeSet<ClusteringElements> rangeSet, QueryOptions options)
+    {
+        assert operator.isSlice() || operator == Operator.EQ;
+        return operator.restrict(rangeSet, bindAndGetClusteringElements(options));
+    }
+
+    private List<ClusteringElements> bindAndGetClusteringElements(QueryOptions options)
+    {
+        switch (columnsExpression.kind())
+        {
+            case SINGLE_COLUMN:
+            case TOKEN:
+                return bindAndGet(options).stream()
```

Comment:
Streams in hot path.

Reply:
You will get that call between zero and 2 times at most for a given query. Considering that token and multi columns are rarely used. Is it really the hot path?

Reply:
Why run the risk :shrug: ?